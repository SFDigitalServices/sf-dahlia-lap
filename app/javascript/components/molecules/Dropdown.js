import React from 'react'
import { find } from 'lodash'
import classNames from 'classnames'

import DropdownMenu from '../molecules/DropdownMenu'
import DropdownMenuMultiSelect from '../molecules/DropdownMenuMultiSelect'

class Dropdown extends React.Component {
  state = { expanded: false, style: { top: 40, left: 0 } }

  componentDidMount () {
    document.addEventListener('mousedown', this.handleClickOutside)
    document.addEventListener('keydown', this.onEscapeHandler)
  }

  componentWillUnmount () {
    document.removeEventListener('mousedown', this.handleClickOutside)
    document.removeEventListener('keydown', this.onEscapeHandler)
  }

  toggleExpand = (e) => {
    this.setState((prevState) => ({ expanded: !prevState.expanded }))
    // We need this to avoid triggering the collapse function from Foundation
    e.stopPropagation()
  }

  componentClickHandler = (e) => {
    // We want to hide the dropdown menu only when we click in the component but outside the menu
    if (this.wrapperRef === e.target) { this.setState({ expanded: false }) }
  }

  onChangeHandler = (value, label) => {
    if (!this.props.multiple) { this.setState({expanded: false}) }
    if (!this.props.multiple) { this.buttonRef.focus() }
    this.props.onChange && this.props.onChange(value, label)
  }

  menu = (classes) => {
    const { items, value, multiple } = this.props
    if (this.state.expanded) {
      if (multiple) {
        return (
          <DropdownMenuMultiSelect
            style={this.state.style}
            onChange={this.onChangeHandler}
            values={value}
            items={items}
            classes={classes} />
        )
      } else {
        return (
          <DropdownMenu
            style={this.state.style}
            onChange={this.onChangeHandler}
            value={value}
            items={items}
            classes={classes} />
        )
      }
    }
  }

  handleClickOutside = (event) => {
    if (this.wrapperRef && !this.wrapperRef.contains(event.target)) {
      this.setState({expanded: false})
    }
  }

 onEscapeHandler = (event) => {
   if (event.keyCode === 27) { this.setState({expanded: false}) }
 }

 render () {
   const {
     prompt,
     items,
     value,
     styles,
     buttonClasses,
     wrapperClasses,
     menuClasses = [],
     disabled
   } = this.props
   const selectedItem = find(items, { value: value })

   return (
     <div
       className={classNames('dropdown', wrapperClasses)}
       onClick={this.componentClickHandler}
       ref={(node) => { this.wrapperRef = node }}
       style={styles}>
       <button
         aria-expanded={this.state.expanded ? 'true' : 'false'}
         onClick={this.toggleExpand}
         ref={(node) => { this.buttonRef = node }}
         className={`button dropdown-button has-icon--right text-align-left ${buttonClasses ? buttonClasses.join(' ') : ''}`}
         type='button'
         disabled={disabled}>
         <span className='ui-icon ui-small'>
           <svg>
             <use xlinkHref='#i-arrow-down' />
           </svg>
         </span>
         {selectedItem ? selectedItem.label : prompt}
       </button>
       <div className='dropdown-menu-wrapper' aria-hidden={this.state.expanded ? 'false' : 'true'} role='menu'>
         {this.menu(menuClasses)}
       </div>
     </div>
   )
 }
}

export default Dropdown
