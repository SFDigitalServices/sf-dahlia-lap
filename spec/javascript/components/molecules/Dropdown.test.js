import React from 'react'

import { shallow } from 'enzyme'
import { components } from 'react-select'

import Dropdown from 'components/molecules/Dropdown'

// Use a string selector in these tests to avoid confusion with the StateManager wrapper that exists in react-select
const selectSelector = 'Select'

const defaultOptionRenderer = ({ label }) => (
  <li>
    <a>{label}</a>
  </li>
)

const defaultToggleRenderer = ({ children, getValue, ...props }) => {
  const val = getValue()[0]
  return (
    <button>
      {val?.label ? val.label : 'placeholder'}
      <components.ValueContainer getValue={getValue} {...props}>
        {children}
      </components.ValueContainer>
    </button>
  )
}

const getWrapper = (propOverrides = {}) =>
  shallow(
    <Dropdown
      renderOption={defaultOptionRenderer}
      renderToggle={defaultToggleRenderer}
      onChange={jest.fn()}
      {...propOverrides}
    />
  )

describe('Dropdown', () => {
  let wrapper
  test('it renders a select', () => {
    wrapper = getWrapper()
    expect(wrapper.dive().find(selectSelector).exists()).toBeTruthy()
  })

  test('it renders with default props correctly', () => {
    wrapper = getWrapper()
    const selectProps = wrapper.dive().find(selectSelector).props()
    expect(selectProps.classNamePrefix).toBeNull()
    expect(selectProps.isDisabled).toBe(false)
    expect(selectProps.options).toEqual([])
    expect(selectProps.value).toBeNull()
    expect(selectProps.formatOptionLabel).toBe(defaultOptionRenderer)
    expect(selectProps.components.ValueContainer).toBe(defaultToggleRenderer)
  })

  test('it passes the passes provided props to the Select component', () => {
    const sampleItems = [
      { value: 'value1', label: 'label1' },
      { value: 'value2', label: 'label2' },
      { value: 'value3', label: 'label3' }
    ]
    const testClassNamePrefix = 'classNamePrefix'
    wrapper = getWrapper({
      items: sampleItems,
      value: 'value2',
      classNamePrefix: testClassNamePrefix,
      disabled: true
    })
    const selectProps = wrapper.dive().find(selectSelector).props()
    expect(selectProps.classNamePrefix).toEqual(testClassNamePrefix)
    expect(selectProps.isDisabled).toBe(true)
    expect(selectProps.value).toEqual(sampleItems[1])
    expect(selectProps.options).toEqual(sampleItems)
  })
})
