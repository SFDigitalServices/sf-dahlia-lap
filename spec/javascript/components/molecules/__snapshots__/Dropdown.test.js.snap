// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown it passes the passes provided props to the Select component 1`] = `
<DocumentFragment>
  <div
    class="dropdown classNamePrefix--is-disabled css-3iigni-container"
  >
    <span
      class="css-1f43avz-a11yText-A11yText"
      id="react-select-4-live-region"
    />
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      class="css-1f43avz-a11yText-A11yText"
    />
    <div
      aria-disabled="true"
      class="classNamePrefix__control classNamePrefix__control--is-disabled css-uuqjpg-Control"
    >
      <button>
        label2
        <div
          class="classNamePrefix__value-container classNamePrefix__value-container--has-value css-1fdsijx-ValueContainer"
        >
          <input
            aria-autocomplete="list"
            aria-expanded="false"
            aria-haspopup="true"
            aria-readonly="true"
            class="css-mohuvp-dummyInput-DummyInput"
            disabled=""
            id="react-select-4-input"
            inputmode="none"
            role="combobox"
            tabindex="0"
            value=""
          />
        </div>
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Dropdown it renders with default props correctly 1`] = `
<DocumentFragment>
  <div
    class="dropdown css-b62m3t-container"
  >
    <span
      class="css-1f43avz-a11yText-A11yText"
      id="react-select-3-live-region"
    />
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      class="css-1f43avz-a11yText-A11yText"
    />
    <div
      class=" css-uuqjpg-Control"
    >
      <button>
        placeholder
        <div
          class=" css-1fdsijx-ValueContainer"
        >
          <div
            class=" css-1jqq78o-placeholder"
            id="react-select-3-placeholder"
          />
          <input
            aria-autocomplete="list"
            aria-describedby="react-select-3-placeholder"
            aria-expanded="false"
            aria-haspopup="true"
            aria-readonly="true"
            class="css-mohuvp-dummyInput-DummyInput"
            id="react-select-3-input"
            inputmode="none"
            role="combobox"
            tabindex="0"
            value=""
          />
        </div>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
